include/lo2s/perf/reader.hpp

```diff
@@ -0,0 +1,368 @@
+        attr_.type = type;
+        attr_.config = config;
+        attr_.config1 = config1;
+    }
+
+    void as_counter();
+    void as_group_leader();
+
+private:
+    std::string name_ = "";
+    std::set<Cpu> cpus_;
+    Availability availability_ = Availability::UNAVAILABLE;
+};
+
+/**
+ * Contains an event parsed from sysfs
+ * @note call on as_sample() after creation to get a valid
+ * event, otherwise the availability will be set to UNAVAILABLE
+ */
+class SysfsEvent : public PredefEvent
+{
+public:
+    using PredefEvent::PredefEvent;
+
+    void as_sample();
+    bool degrade_percision();
+};
+
+class TimeEvent : public PerfEvent
+{
+public:
+    TimeEvent(bool enable_on_exec, uint64_t addr);
+};
+
+/**
+ * Contains an event that is addressable via name
+ * @note Call on as_syscall() after creation if needed
+ */
+class TracepointEvent : public PerfEvent
+{
+public:
+    class ParseError : public std::runtime_error
+    {
+    public:
+        ParseError(const std::string& what) : std::runtime_error(what)
+        {
+        }
+
+        ParseError(const std::string& what, int error_code);
+    };
+
+    TracepointEvent() : id_(-1)
+    {
+    }
+    TracepointEvent(bool enable_on_exec, int event_id);
+    TracepointEvent(bool enable_on_exec, const std::string& name);
+
+    friend bool operator==(const TracepointEvent& lhs, const TracepointEvent& rhs)
+    {
+        return lhs.id_ == rhs.id_;
+    }
+
+    friend bool operator<(const TracepointEvent& lhs, const TracepointEvent& rhs)
+    {
+        return lhs.id_ < rhs.id_;
+    }
+
+    friend bool operator>(const TracepointEvent& lhs, const TracepointEvent& rhs)
+    {
+        return lhs.id_ > rhs.id_;
+    }
+
+    void parse_format();
+
+    void as_syscall()
+    {
+        attr_.sample_type |= PERF_SAMPLE_IDENTIFIER;
+    }
+
+    const auto& fields() const
+    {
+        return fields_;
+    }
+
+    const auto& common_fields() const
+    {
+        return common_fields_;
+    }
+
+    const auto& field(const std::string& name) const
+    {
+        for (const auto& field : fields())
+        {
+            if (field.name() == name)
+            {
+                return field;
+            }
+        }
+        throw std::out_of_range("field not found");
+    }
+
+    int id()
+    {
+        return id_;
+    }
+
+    std::string name() const
+    {
+        return name_;
+    }
+
+    static std::vector<std::string> get_tracepoint_event_names();
+
+private:
+    void parse_format_line(const std::string& line);
+
+    const static std::filesystem::path base_path_;
+    int id_;
+    std::string name_;
+    std::vector<tracepoint::EventField> common_fields_;
+    std::vector<tracepoint::EventField> fields_;
+};
+
+/**
+ * Contains a raw event for testing
+ */
+class RawEvent : public PerfEvent
+{
+public:
+    RawEvent(bool enable_on_exec);
+};
+
+/**
+ * Contains an opened instance of PerfEvent.
+ * Use PerfEvent.open() to construct an object
+ */
+class PerfEventInstance
```

Comment:
Also, I have the feeling, that it might be good to have a template argument to this class, which is the actual event used, e.g., `EventGuard<TracepointEvent> event_;`.

Reply:
But the type of `Event` in `EventGuard` isn't being checked anywhere. Is there even a use for this information in `EventGuard`?